Improve (de)serialization and validation capabilities (#653)
python-attrs/attrs: docs/extending.rst

    ...     x = typed(int, default=1, init=False)
    >>> attr.fields(C).x.metadata[MY_TYPE_METADATA]
    <class 'int'>

> 

Reviewer: This is for a reference from the `attr.s` docstring below.

```suggestion

.. _transform-fields:
```

Automatic field transformation and modification
------------------------------------------------

Attrs allows you to automatically modify or transform the class' fields while the class is being created.
You do this by passing a *field_transformer* hook to :func:`attr.define()` (and its friends).
Its main purpose is to automatically add converters to attributes based on their type to aid the development of API clients and other typed data loaders.

| This hook must have the signature:

> | :code:`your_hook(cls: type, fields: List[Attribute]) -> List[Attribute]`

Author: The pipes just enforce manual line breaks. Your solution seems a lot better. :)

- *cls* is your class right *before* it is being converted into an attrs class.
  This means it does not yet have the ``__attrs_attrs__`` attribute.

> - *fields* is a list of all :class:`attr.Attribute` instances that will later be set to ``__attrs_attrs__``.  You can modify these attributes any way you want:

Reviewer: ```suggestion
- *fields* is a list of all :class:`attr.Attribute` instances that will later be set to ``__attrs_attrs__``. 
  You can modify these attributes any way you want:
```

   ...
   >>> @attr.frozen(field_transformer=drop_floats)
   ... class Data:
   ...     a: int
   ...     b: float
   ...     c: str
   ...
   >>> Data(42, "spam")
   Data(a=42, c='spam')

A more realistic example would be to automatically convert data that you, e.g., load from JSON:

.. doctest::

   >>> from datetime import datetime
   >>>
   >>> def auto_convert(cls, fields):
   ...     results = []
   ...     for field in fields:
   ...         if field.converter is not None:
   ...             results.append(field)
   ...             continue
   ...         if field.type in {datetime, 'datetime'}:
   ...             converter = (lambda d: datetime.fromisoformat(d) if isinstance(d, str) else d)
   ...         else:
   ...             converter = None
   ...         results.append(field._assoc(converter=converter))
   ...     return results
   ...
   >>> @attr.frozen(field_transformer=auto_convert)
   ... class Data:
   ...     a: int
   ...     b: str
   ...     c: datetime
   ...
   >>> from_json = {"a": 3, "b": "spam", "c": "2020-05-04T13:37:00"}
   >>> Data(**from_json)  # ****
   Data(a=3, b='spam', c=datetime.datetime(2020, 5, 4, 13, 37))

> Change value serialization in ``asdict()``
> ------------------------------------------

Reviewer: I think Customize is better here:

```suggestion
Customize Value Serialization in ``asdict()``
---------------------------------------------
```

> Attrs allows you to serialize instances of attrs classes to dicts using the :func:`attr.asdict()` function.

Reviewer: ```suggestion
``attrs`` allows you to serialize instances of ``attrs`` classes to dicts using the `attr.asdict` function.
```

However, the result can not always be serialized since most data types will remain as they are:

.. doctest::

   >>> import json
   >>> import datetime
   >>>
   >>> @attr.frozen
   ... class Data:
   ...    dt: datetime.datetime
   ...
   >>> data = attr.asdict(Data(datetime.datetime(2020, 5, 4, 13, 37)))
   >>> data
   {'dt': datetime.datetime(2020, 5, 4, 13, 37)}
   >>> json.dumps(data)
   Traceback (most recent call last):
     ...
   TypeError: Object of type datetime is not JSON serializable

> To help you with this, :attr:`attr.asdict()` allows you to pass a *value_serializer* hook.  It has the signature :code:`your_hook(inst: type, field: Attribute, value: Any) -> Any`:

Reviewer: ```suggestion
To help you with this, `attr.asdict` allows you to pass a *value_serializer* hook.
It has the signature

.. function:: your_hook(inst: type, field: attr.Attribute, value: typing.Any) -> typing.Any
   :noindex:
```
